[hingen] iP

hingen opened the pull request:
## New Super Duke Pro Definitive Edition 4K Remastered :open_mouth::open_mouth::open_mouth:
> “_Your mind is for having ideas, not holding them._” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

Introducing **Duke**! The manager of tasks, tracker of progress!
The less time you spend remembering, the more time you spend doing. Organize your mind with **Duke**!

## Features
* Track tasks
* Text-based interface
* Available for both CLI and GUI

## Building Duke from Source
1. Clone the repo: `git clone git@github.com:hingen/ip.git`
2. Navigate to the project root directory: `cd {project_root}`
3. Build with gradle: `./gradlew shadowjar`
4. `duke.jar` can be found in `{project_root}/build/libs/duke.jar`

## Running Duke
```bash
# launch GUI version
java -jar duke.jar

# launch CLI version
java -jar duke.jar --cli
```

## Roadmap
- [X] Managing tasks
- [ ] Reminders
- [ ] Mobile app

## Practice Coding with Duke
**Duke** is **open source**!!!
If you're a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public static void main(String[] args) {
    if (shouldLaunchCli(args)) {
        DukeCli.run();
    } else {
        DukeGui.run(args);
    }
}
```

Commit:
Add Gradle support

Commit:
Move Duke.java to duke package

Every class should be part of some package.

Let's move Duke.java into the duke package.

Since Duke.java contains the main(String[]) method, it should be placed
in the package at the top of the package hierarchy which is duke.

Commit:
Duke class: change format of greeting message

The greeting message does not comply with the intended format for all
messages produced by the bot.

Let's create a method printMessage(String) that will take in an
unformatted message and print it out in the bot's message format. The
greeting message is then printed using this method call.

The bot is expected to print other messages that must also comply with
the message format. Using a method that will handle the formatting will
reduce code duplication.

Commit:
Add Command.java

The user can input different commands but all commands share common
functionalities. That is all commands produce a response message base on
the user's input.

Let's add a Command interface.

A Command interface allows for reduce code duplication via polymorphism.
It also allows for better structuring by giving each command type it's
own class which would contain the logic for how the command is to be
run.

Commit:
Add EchoCommand.java

The user's input needs to be echoed to the user.

Let's add an EchoCommand class which implements the Command interface.

An EchoCommand class would encapsulate the logic for the echo command
type and can be utilised through the Command interface.

Commit:
Add ByeCommand.java

A farewell message needs to be displayed when the user inputs bye.

Let's add a ByeCommand class which implements the Command interface.

A ByeCommand class would encapsulate the logic for the bye command type
and can be utilised through the Command interface.

Commit:
Duke class: change to support user input with echo and bye command

User's need to be able to input echo and bye commands.

Let's change main(String[]) to have a loop that takes user inputs and
for each input, run the appropriate command and display the resulting
message. Exit the loop when the bye command is inputted.

The code is written such that future modifications can be easily made to
account for other command types.

Commit:
Fix extra new line in the bot's response messages

Commit:
Add javadocs for Duke.java

All public methods and classes should have descriptive comments.

Let's add javadocs for Duke.java class.

javadocs are shown by the IDE when hovering an item which is documented.
It can also be used for generating documentation later.

Commit:
ByeCommand class: remove empty line

Commit:
EchoCommand class: remove empty line

Commit:
Command class: change run method to support modifying user's task list

Certain commands must be able to modify the user's task list.

Let's
* change the Command::run method to also take the user's current task
  list as input
* create a CommandResponse class which contains the response message and
  updated tasks list resulting from a command
* change Command::run to return a CommandResponse object

As the chatbot will have to manage the user's task list, it is expected
that almost all future commands will have to modify the user's task
list, hence Command::run should be change to reflect this.

Commit:
Add support for add command

The user needs to add tasks to their task list.

Let's add an AddCommand class which implements the Command interface and
have the command run whenever the user inputs anything other than "bye".

An AddCommand class would encapsulate the logic for the add command type
and can be utilised through the Command interface.

Commit:
Add support for list command

The user needs to list out all tasks in their task list.

Let's add a ListCommand class which implements the Command interface and
have the command run whenever the user inputs "list".

A ListCommand class would encapsulate the logic for the list command
type and can be utilised through the Command interface.

Commit:
Add support for task done status

Tasks need to have a done status for user's to track which task they
have and have not done.

Let's,
* add a Task class to store a task's done status and description
* change the current implementation of tasks stored in a List<String> to
  List<Task>

A Task class allows for packaging of all information related to a
specific task into a single location.

Commit:
Add support for printing a string containing Task object info

The info related to a task should be printed out in a standardised
format.

Let's,
* add a toString() method to Task class which returns a string
  containing the task's info in a standardised format
* change ListCommand::run(String, List<Task>) to utilise
  Task::toString()

To standardise the way task info is presented, a method for getting a
string in said format should exist in the Task class.

Commit:
Add support for marking task as done

The user needs to be able to mark a task as done.

Let's,
* add a MarkCommand class containing the logic for marking a task as
  done
* change Duke::main(String[]) to run the mark command when the user
  enters an input that begins with mark

A MarkCommand class would encapsulate the logic for the mark command
type and can be utilised through the Command interface.

Commit:
Add support for marking task as not done

The user needs t o be able to mark a task as not done.

Let's,
* add an UnmarkCommand class containing the logic for marking a task as
  not done
* change Duke::main(String[]) to run the unmark command when the user
  enters an input that begins with unmark

An UnmarkCommand class would encapsulate the logic for the unmark
command type and can be utilised through the Command interface.

Commit:
Fix missing new line in javadocs

Commit:
Add support for ToDo task

The user needs to be able to add different types of task, one of them
being the To-Do task.

Let's,
* add a ToDoTask class which inherits Task and contains any logic
  related to the To-Do task type
* add a ToDoCommand class which handles To-Do commands
* change Duke class to run the ToDoCommand::run(String, List<Task>)
  method whenever the user's input starts with "todo"

Commit:
Add support for deadline task

The user needs to be able to add deadline type task.

Let's,
* add a DeadlineTask class which inherits Task and contains any logic
  related to the deadline task type
* add a DeadlineCommand class which handles deadline commands
* change Duke class to run the DeadlineCommand::run(String, List<Task>)
  method whenever the user's input starts with "deadline"

Commit:
Add support for event task

The user needs to be able to add event type task.

Let's,
* add a EventTask class which inehrits from Task and contains any logic
  related to the event task type
* add a EventCommand class which handles event commands
* change Duke class to run the EventCommand::run(String, List<Task>)
  method whenever the user's input starts with "event"

Commit:
Rename DeadlineTask, EventTask, ToDoTask to Deadline, Event, ToDo

Commit:
Task.java class: remove unused methods

Commit:
Add support for handling errors in user's input

Users may input invalid commands which need to be handled.

Let's,
* add a DukeException class for handling errors in the user's input
* change Command::run(String, List<Task>) to allow for the throwing of a
  DukeException whenever there an error results from the user's input
* change Duke::main(String[]) to print out the error message

Commit:
Remove EchoCommand.java

EchoCommand class was previously used as a placeholder. It is no longer
needed.

Commit:
Command class: add missing javadocs for run(String, List<Task>)

Commit:
UnmarkCommand class: add handling of input errors

User input for unmark command might be missing an index or the index is
not an integer or the index is out of bounds.

Let's throw a DukeException with the error described in the exception
message when such errors occur.

Commit:
MarkCommand class: add handling of input errors

User input for mark command might be missing an index or the index is
not an integer or the index is out of bounds.

Let's throw a DukeException with the error described in the exception
message when such errors occur.

Commit:
DeadlineCommand class: add handling of input errors

User input for deadline command might be missing a ' /by '.

Let's throw a DukeException with the error described in the exception
message when such an error occurs.

Commit:
EventCommand class: add handling of input errors

User input for event command might be missing a ' /from ' and/or
' /to '.

Let's throw a DukeException with the error described in the exception
message when such an error occurs.

Commit:
Fix missing punctuation in javadocs

Commit:
Duke class: add handling of unknown command

User input might not be a known command.

Let's print an error message when such an error occurs.

Commit:
Fix missing punctuation in error messages

Commit:
Fix unwanted removal of 'deadline' from deadline description

Commit:
Fix unwanted deletion of 'event' from event description

Commit:
Fix deadline command failing with ' /by ' not found even when it exist

Commit:
Fix event command failing with ' /to ' not found even when it exist

Commit:
ToDoCommand class: add handling of input errors

User input for todo command might have an empty description.

Let's throw a DukeException with the error described in the exception
message when such an error occurs.

Commit:
DeadlineCommand class: add handling of empty description in input

User input for deadline command might have an empty description.

Let's throw a DukeException when such an error occurs.

Commit:
EventCommand class: add handling of empty description in input

User input for event command might have an empty description.

Let's throw a DukeException when such an error occurs.

Commit:
Fix empty whitespace in deadline and event

Commit:
DeadlineCommand class: fix outdated javadocs

Commit:
EventCommand class: fix outdated javadocs

Commit:
Add support for deleting task

The user needs to be able to delete tasks.

Let's,
* add a DeleteCommand class which contains the logic for deleting tasks
* change Duke class to run the DeleteCommand::run(String, List<Task>)
  method whenever the user's input starts with "delete"

Commit:
Fix chatbot not exiting when bye command is inputted with arguments

Commit:
Duke class: move exit condition into a static method

SLAP property is violated by having the exit condition be in the
Duke::runInputLoop() method.

Let's move the exit condition into a static method to maintain SLAP
property.

SLAP makes the code more readable.

Commit:
DeleteCommand class: split up run(String, List<Task>) method

The run(String, List<Task>) method consist of many sections.

Let's split the run(String, List<Task>) method in different methods.

This will improve code readability.

Commit:
MarkCommand and UnmarkCommand classes: extract a parent class

MarkCommand and UnmarkCommand have very similar logic.

Let's extract the common behaviour of both classes into a parent class
SetDoneCommand which both classes can then extend.

Commit:
ToDoCommand class: split createTask(String) method

The createTask(String) method consist of an input validation section and
task creation section.

Let's split the createTask(String) method into different methods.

This will improve code readability.

Commit:
DeadlineCommand class: split createTask(String) method

The createTask(String) method consist of an input validation section and
task creation section.

Let's split the method into different methods.

This will improve code readability.

Commit:
EventCommand class: split createTask(String) method

The createTask(String) method consist of an input validation section and
task creation section.

Let's split the method into different methods.

This will improve code readability.

Commit:
DukeException class: fix compile error due to emoticon

Commit:
EventCommand class: combine unspecified start or end date/time errors

There is no need to be that specific with the missing date range error.

Let's combine the unspecified start or end date/time error into a
missing unspecified date range error.

This will allow the code to be simpler.

Commit:
Fix mark command success message being used for unmark command success;

Commit:
SetDoneCommand class: remove commented out code

Commit:
DeleteCommand class: shorten non-existent task error

The non-existent task error is too long.

Let's shorten it.

Short and concise errors are easier to read.

Commit:
SetCommand class: shorten non-existent task error

The non-existent task error is too long.

Let's shorten it.

Short and concise errors are easier to read.

Commit:
Add testing

Commit:
Duke class: fix scanner not closed

Commit:
Duke class: fix violation of coding standard by case indent

Commit:
Replace List<Task> with TaskList

There are operations that a task list should have which does not exist
in List such as a toString that is specific to TaskList.

Let's create a TaskList class and replace all usage of List<Task> with
TaskList.

This allows us to add task list specific operations as methods to
TaskList.

Commit:
Add support for saving and loading task list to storage

The task list needs to be persistent and thus it needs to be written to
and read from storage.

Let's,
* create a Storage interface for our Storage logic
* create a FileStorage class for our reading and writing to file logic
* change TaskList to read and write to storage whenever the task list is
  updated by using a FileStorage object

Commit:
Add restriction to event and deadline to require date and time

The start and end of an event and the cutoff of a deadline should be
dates and time.

Let's,
* change Event to require start and end to be LocalDateTime objects
* change Deadline to require cutoff to be a LocalDateTime object
* change DeadlineCommand to only accept cutoff inputs that are of a
  specific format and are valid date and time
* change EventCommand to only accept start and end inputs that are of a
  specific format and are valid date and time

Commit:
Merge branch 'branch-Level-7'

Commit:
Merge branch 'branch-Level-8'

Commit:
Fix Deadline and Event createFromStorage method not parsing date time

Commit:
Extract logic for printing messages to Ui.java class

Commit:
Extract logic for parsing user input to Parser.java class

Commit:
Merge remote-tracking branch 'remotes/origin/add-gradle-support'

Commit:
Fix mark and unmark commands not updatng saved task list

Commit:
Add JUnit tests for BooleanUtils

Commit:
Add JUnit tests for LocalDateTimeUtils

Commit:
Add JUnit tests for Ui

Commit:
Customise output messages

Commit:
ListCommand class: add specific output for when task list is empty

Commit:
Update outdated JavaDocs

Commit:
Fix coding standard violations

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Add support for find command

The user needs to be able to search for commands that contain a specific
keyword or keyphrase.

Let's,
* add a FindCommand class that contains the logic for filtering a task
  list by a specific keyword or keyphrase
* change Parser class to use the FindCommand::run method when the user
  enters the find command

Commit:
Merge branch 'branch-Level-9'

Commit:
Add JUnit tests for ToDo

Commit:
Fix coding standard violations

Commit:
Extract command-line interface logic into Cli class

There might be several different types of UI that the app the must
support.

Let's,
* change Ui class to be an interface
* extract the command-line interface logic into the Cli class and have
  it implement the Ui interface

Commit:
Fix find command not listing sequential index

Commit:
Revert "Fix find command not listing sequential index"

This reverts commit 3b914dc771f86bc8236ecfabbdac4e3d4c8b59c7.

Commit:
Fix minor code format issues

Commit:
Refactor code base to support JavaFX GUI

Commit:
Add support for GUI

Commit:
Fix text-ui-test failing due to app not running in CLI

Commit:
Merge branch 'branch-Level-10'

Commit:
Change styling of GUI

Commit:
Merge branch 'branch-Level-10'

Commit:
Add checkstyle

Commit:
Fix issues detected by checkstyle

Commit:
Change styling of dialog box

Commit:
.gitignore: change to ignore /.vscode/ directory

Commit:
DialogBox class: remove unuse import

Commit:
Refactor classes in duke.utils to improve code quality

Commit:
Refactor Duke CLI classes to improve code quality

Commit:
Gui class: replace Magic literal for stage width and height

Stage width and height are currently being set using magic numbers which
are detrimental to code readability.

Let's replace the magic numbers with 2 private symbolic constants,
STAGE_WIDTH and STAGE_HEIGHT.

Using symbolic constants makes the intention of the constants more
explicit and thus improves code readability. The constants are kept
private as they are not needed by other classes.

Commit:
Fix missing space between brackets

Commit:
Refactor Duke GUI classes to improve code quality

tricixg commented on src/main/java/duke/command/ToDoCommand.java:
```diff
@@ -0,0 +1,47 @@
+ * Represents an add to-do command for adding a to-do to a task list.
+ */
+public class ToDoCommand extends AddCommand {
+    /**
```
good format of javadocs

tricixg commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,87 @@
+     * @param args Data about the event that was loaded from storage.
+     * @return The Event object created using the data loaded from storage.
+     * @throws DukeException Indicates missing data or incorrect data type or format in args.
+     */
```
good that there is a checker for stored values. easy to understand

tricixg commented on src/main/java/duke/ui/cli/Cli.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    @Override
+    public void print(String message) {
```
can use variable to store dividers instead of println

hingen replied:
Thanks for the suggestion! Will definitely do so.

tricixg reviewed (commented):
LGTM. overall very well documented and follows good coding style

Commit:
Refactor classes in duke.task package to improve code quality

Commit:
Cli class: extract message border to constant

The message border for the CLI version of Duke is a magic literal and is
duplicated (one for the start of the message, and one for the end of the
message).

Let's extract the message border into a private constant within the Cli
class.

Since message border is unchanging, store it as a constant.

There is no need to access it outside of the class hence give it the
private access modifier.

By having the message border literal stored in a variable, there is no
longer a need to duplicate the border.

Commit:
Task class: fix incorrect access modifier order

Commit:
Refactor duke.io classes to improve code quality

Commit:
Standardise return comments

Commit:
text-ui-test/runtest.sh script: fix classes not rebuilding after change

Commit:
Cli class: fix missing new line when printing message

Commit:
Refactor classes in duke.command package to improve code quality

Commit:
Refactor duke.task classes to improve code quality

Commit:
Add assertions

hingen committed:
Merge pull request #1 from hingen/branch-A-CodeQuality

Improve code quality

nusE0726844 reviewed (commented):
I like the readability of your code

Commit:
Merge branch 'master' into branch-A-Assertions

Commit:
Add assertions for easier error detection

hingen committed:
Merge pull request #2 from hingen/branch-A-Assertions

Add assertions

Commit:
BooleanUtils class: change isBooleanStr(String) to check for null

Commit:
TaskList class: fix missing exception JavaDocs for update(int, Task)

Commit:
Fix JavaFX warning due to FXML version mismatch

Commit:
runtest.bat: change to use gradle for building

Commit:
checkstyle.xml: fix import order not matching for org.* imports

Commit:
Fix code style errors

Commit:
Add GitHub workflow for checking code upon push or PR

Commit:
Add support for archiving task list

The user may have completed all tasks in the task list but do not want
to lose the data on the task by deleting all of them.

Let's
* add an ArchiveCommand class that handles archiving the user's task
  list to a file and clears the task list
* change Parser class to run ArchiveCommand::run(String, TaskList)
  whenever the user's input begins with "archive"

By encapsulating the logic for archiving the task list into the
ArchiveCommand, we can make it easier for developers to find said logic.
It also allows complies with the single-responsibility principle.

hingen committed:
Merge pull request #3 from hingen/branch-C-Archive

Add support for archiving task list

Commit:
Extract magic number array indexes to symbolic constant

There are multiple occurrences of magic numbers being use to index
arrays.

Let's extract these magic numbers to private constants within their
respective classes.

Using symbolic constants makes it clear what element are being indexed
thus improving code readability.

Commit:
ToDo class: Fix unused EXPECTED_ARG_COUNT

Commit:
Fix app crash when archiving to "./"

Commit:
Gui class: set stage title to "Duke"

Commit:
Storage class: fix incorrect JavaDocs convention

Commit:
Add product website

Commit:
Add section linking for user guide

Commit:
Add dracula theme for user guide

Commit:
Fix misleading phrasings in user guide

Commit:
Add screenshot to user guide
